src/GameLoop.tsx

```diff
@@ -55,223 +59,220 @@ const stateMachine = createMachine({
-        return;
-    }
-    const board = boardPhysics.boardCellMatrix;
-    const r = playerPhysics.pos[0];
-    const c = playerPhysics.pos[1];
-    const areTargetSpacesEmpty = (dr, dc) =>
-        playerPhysics.adjustedCells.every((cell) => {
-            return board[cell.r + dr][cell.c + dc].char === EMPTY;
-        });
-    if (keyCode === 37) {
-        // Left
-        if (
-            playerPhysics.isInCBounds(
-                playerPhysics.getAdjustedLeftmostC() - 1,
-            ) &&
-            // Ensure blocks don't cross over to ground higher than it, regarding interpolation.
-            (!ENABLE_SMOOTH_FALL ||
-                playerPhysics.isInRBounds(
-                    playerPhysics.getAdjustedBottomR() +
-                        Math.ceil(interp.val / interpMax),
-                )) &&
-            areTargetSpacesEmpty(
-                Math.ceil(ENABLE_SMOOTH_FALL ? interp.val / interpMax : 0),
-                -1,
-            )
-        ) {
-            playerPhysics.setPos(r, c - 1);
-            playerPhysics.hasMoved = true;
-        }
-    } else if (keyCode === 39) {
-        // Right
-        if (
-            playerPhysics.isInCBounds(
-                playerPhysics.getAdjustedRightmostC() + 1,
-            ) &&
-            // Ensure blocks don't cross over to ground higher than it, regarding interpolation.
-            (!ENABLE_SMOOTH_FALL ||
-                playerPhysics.isInRBounds(
-                    playerPhysics.getAdjustedBottomR() +
-                        Math.ceil(interp.val / interpMax),
-                )) &&
-            areTargetSpacesEmpty(
-                Math.ceil(ENABLE_SMOOTH_FALL ? interp.val / interpMax : 0),
-                1,
-            )
-        ) {
-            playerPhysics.setPos(r, c + 1);
-            playerPhysics.hasMoved = true;
-        }
-    } else if (keyCode === 40) {
-        // Down
-        if (repeat) {
-            // TODO: Handle repeated downkey.
-        }
-        if (
-            playerPhysics.getAdjustedBottomR() + 1 < BOARD_ROWS &&
-            areTargetSpacesEmpty(1, 0)
-        ) {
-            if (ENABLE_SMOOTH_FALL) {
-                interp.val += interpRate * interpKeydownMult;
-            } else {
-                playerPhysics.setPos(r + 1, c);
-            }
+export function GameLoop() {
+    const [boardPhysics, _setBoardPhysics] = useState(
+        new BoardPhysics(BOARD_ROWS, BOARD_COLS),
+    );
+    const [_boardCellMatrix, setBoardCellMatrix] = useState(
+        boardPhysics.boardCellMatrix,
+    );
+
+    const [playerPhysics, _setPlayerPhysics] = useState(
+        new PlayerPhysics(boardPhysics),
+    );
+    const [_adjustedCells, setAdjustedCells] = useState(
+        playerPhysics.adjustedCells,
+    );
+    const [isPlayerVisible, setPlayerVisibility] = useState(true);
+
+    useEffect(() => {
+        globalThis.requestAnimationFrame(loop);
+        globalThis.addEventListener("keydown", updatePlayerPos);
```

Comment:
Should this return a `removeEventListener`?  I thought I had saw that in an earlier commit.

Reply:
In normal uses of `useEffect`, yes, but supplying a `[]` as the 2nd argument to `useEffect` tells it to run only once, so a return statement will never be  called. It works as intended

Reply:
You can find that information [here](https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects)

Reply:
Yeah but it doesn't compose as well.  If we ever add a start screen or leaderboard page or something, we'll want to stop it.

Reply:
Let's deal with that when we get there